### PR TITLE
Indirect -  Save Custom Detector Grouping to XML

### LINKED ISSUE #29984: Indirect Reduction Save Custom Detector Grouping
**Original reporter:** @SpencerHowells

### Description
Indirect Data Reduction needs the ability to save a custom grouping of detectors to a `.map` file in order for it to be easily loaded back into the interface.

This will require:
- Adding a button next to the Custom Grouping option that will save the grouping when clicked.
- Converting the custom group string into a `GroupingWorkspace`
- Modification of the `SaveDetectorsGrouping` algorithm to allow the saving of `.map` files.

### Platforms affected
All

### PR DESCRIPTION
**Pull request by robertapplin**
**Report to:** @SpencerHowells

**Description of work.**
This PR adds:
- The ability to create a grouping workspace using a custom grouping string. To do this `CreateGroupingWorkspace` now has the `CustomGroupingString` property. This can create a grouping workspace based on a string such as `3+4+5,6-10,11:15,16`. This is the same syntax used for the `GroupDetectors` algorithm.
- The ability to save a custom grouping string from the Indirect Data reduction interface to an `.xml` file.
- Added the ability to load `.xml` grouping files into indirect data reduction.

It was decided not to add the ability to save detector groupings to a `.map` file. It is better to use `.xml` because:
- It is the standard within Mantid, and is therefore better tested and used by many scientific areas.
- It can do everything a `.map` file can do (and is arguably a simpler format when you get used to it).
- `.map` files are *not* being used in external programs by Indirect, and so `.xml` are just as useful.
- Better documentation on `.xml` format. Used widely in many different algorithms.

**To test:**
1. Indirect Data Reduction interface
2. Select `IRIS` and `graphite` as analyser.
2. Select `Custom` grouping in the grouping combobox
3. Enter a grouping string in the line edit. Try `3+4,6-25,30-50,51:53`
4. Click `Save Grouping`
5. Choose a file name and make sure it ends with `.xml`. Save it.
6. Open the newly created file.
7. Should look like this file:
[iris_detector_grouping_test.zip](https://github.com/mantidproject/mantid/files/5688548/iris_detector_grouping_test.zip)

8. Load this file into Data Reduction from the `File` grouping option.
9. Perform a reduction for `26176` run
10. Show data for the created workspace. It should have 6 spectra.

Closes #29984

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Comment by Harrietbrown** [comment]
The save type from the algorithm is an xml file, not a map
My saved output does not look like yours
mine:
image
yours:
image

**Comment by robertapplin** [comment]
Sorry @Harrietbrown I forgot to update the testing instructions. It has been decided to save the groupings in `.xml` files instead of `.map` files. The testing instructions should be up-to-date now

**Review by Harrietbrown — CHANGES_REQUESTED** [review]
The algorithm can accept strigns like `1+2+3+4` but the indirect interface box does not allow you to type that in (I can't copy and past the string you used as an example.)

**Review by Harrietbrown — DISMISSED** [review]
PR still functions as expected with changes from review. Good Job

**Comment by martyngigg** [comment]
We have a merge conflict now that one of the other indirect PRs has been merged I'm afraid.

**Comment by robertapplin** [comment]
Thanks @martyngigg , it should be fixed now

File: qt/scientific_interfaces/Indirect/IndirectTab.h [hunk 1]
```diff
@@ -88,16 +88,6 @@ class MANTIDQT_INDIRECT_DLL IndirectTab : public QObject, public IPyRunner {
 
   void displayWarning(std::string const &message);
 
-public slots:
```
Comment (Harrietbrown) [inline qt/scientific_interfaces/Indirect/IndirectTab.h]: Should this have gone in?

Comment (robertapplin) [reply qt/scientific_interfaces/Indirect/IndirectTab.h]: I don't know why this is showing as being removed. It should still be there. If you look at `IndirectTab` in the Files Changed it seems to be fine

Comment (Harrietbrown) [reply qt/scientific_interfaces/Indirect/IndirectTab.h]: Strange, perhaps some mix up during the rebase made github think it ahd been changed. it is fine now.

**Comment by Harrietbrown** [comment]
Now when I try a custom grouping of `2+3+4,6-25,30-50,51:53` it causes a crash.

**Comment by robertapplin** [comment]
> Now when I try a custom grouping of `2+3+4,6-25,30-50,51:53` it causes a crash.

Thanks for spotting this, `validateInputs` fails because Detector ID 2 doesn't exist for this instrument. I have now fixed this. There should now be a warning message in the logger saying that this detector ID is invalid.

**Review by Harrietbrown — APPROVED** [review]
With the error catching for invalid inputs the xml saving is working great, the output looks right and the documentation propperly describes how the grouping string should be formatted.
Good Job.

**Pull request merged**